src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -681,54 +681,51 @@ public static bool StartsWithPrefix(string prefix, string key)
                 throw new ArgumentNullException(nameof(key));
             }
 
-            if (StringComparer.OrdinalIgnoreCase.Equals(key, prefix))
+            if (prefix.Length == 0)
             {
+                // Everything is prefixed by the empty string
                 return true;
             }
 
-            if (key.Length <= prefix.Length)
+            if (key.Length < prefix.Length)
             {
                 return false;
             }
 
+            var subKeyIndex = 0;
             if (!key.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
             {
-                if (key.StartsWith("[", StringComparison.OrdinalIgnoreCase))
+                if (key[0] == '[')
                 {
-                    var subKey = key.Substring(key.IndexOf('.') + 1);
+                    subKeyIndex = key.IndexOf('.') + 1;
```

Comment:
Looking into this more with @dougbu all of this code is actually incorrect. This implies that `[0].productId` is a match for `productId`. We're going to fix the behavior of this method and change the code that relies on this behavior (`ClearValidationState`) to do the right thing in isolation